private/buf/cmd/buf/command/curl/curl.go

```diff
@@ -0,0 +1,831 @@
+	endpointURL, service, method, baseURL, err := verifyEndpointURL(container.Arg(0))
+	if err != nil {
+		return err
+	}
+	isSecure := endpointURL.Scheme == "https"
+	if err := f.validate(isSecure); err != nil {
+		return err
+	}
+
+	var clientOptions []connect.ClientOption
+	switch f.Protocol {
+	case protocolGRPC:
+		clientOptions = []connect.ClientOption{connect.WithGRPC()}
+	case protocolGRPCWeb:
+		clientOptions = []connect.ClientOption{connect.WithGRPCWeb()}
+	}
+	if f.Protocol != protocolGRPC {
+		// The transport will log trailers to the verbose printer. But if
+		// we're not using standard grpc protocol, trailers are actually encoded
+		// in an end-of-stream message for streaming calls. So this interceptor
+		// will print the trailers for streaming calls when the response stream
+		// is drained.
+		clientOptions = append(clientOptions, connect.WithInterceptors(traceTrailersInterceptor{printer: container.VerbosePrinter()}))
+	}
+
+	dataSource := "(argument)"
+	var dataFileReference string
+	if strings.HasPrefix(f.Data, "@") {
+		dataFileReference = strings.TrimPrefix(f.Data, "@")
+		if dataFileReference == "-" {
+			dataSource = "(stdin)"
+		} else {
+			dataSource = dataFileReference
+			if absFile, err := filepath.Abs(dataFileReference); err == nil {
+				dataFileReference = absFile
+			}
+		}
+	}
+	requestHeaders, dataReader, err := loadHeaders(f.Headers, dataFileReference, nil)
+	if err != nil {
+		return err
+	}
+	if len(requestHeaders.Values("user-agent")) == 0 {
+		userAgent := f.UserAgent
+		if userAgent == "" {
+			userAgent = defaultUserAgent(f.Protocol)
+		}
+		requestHeaders.Set("user-agent", userAgent)
+	}
+	if dataReader == nil {
+		if dataFileReference == "-" {
+			dataReader = os.Stdin
+		} else if dataFileReference != "" {
+			f, err := os.Open(dataFileReference)
+			if err != nil {
+				return errorHasFilename(err, dataFileReference)
+			}
+			dataReader = f
+		} else if f.Data != "" {
+			dataReader = io.NopCloser(strings.NewReader(f.Data))
+		}
+		// dataReader is left nil when nothing specified on command-line
+	}
+	defer func() {
+		if dataReader != nil {
+			err = multierr.Append(err, dataReader.Close())
+		}
+	}()
+
+	transport, err := makeHTTPClient(f, isSecure, getAuthority(endpointURL, requestHeaders), container.VerbosePrinter())
+	if err != nil {
+		return err
+	}
+
+	output := container.Stdout()
+	if f.Output != "" {
+		output, err = os.Create(f.Output)
+		if err != nil {
+			return errorHasFilename(err, f.Output)
+		}
+	}
+
+	var res resolver
+	closeRes := func() {}
+
+	if f.Reflect {
+		reflectBaseURL := baseURL
+		if f.ReflectBaseURL != "" {
+			reflectBaseURL = f.ReflectBaseURL
+		}
+		reflectHeaders, _, err := loadHeaders(f.ReflectHeaders, "", requestHeaders)
+		if err != nil {
+			return err
+		}
+		res, closeRes = resolverFromReflection(ctx, transport, clientOptions, reflectBaseURL, f.ReflectVersion, reflectHeaders, container.VerbosePrinter())
+		defer closeRes()
```

Comment:
this defer statement, paired with the hard call of `closeRes()` [below](https://github.com/bufbuild/buf/pull/1698/files#diff-05c240e5d4191f595d5b747c112c314cf14ee4d77505cee99aac9500eed39527R717) resets the stream twice, It LOOKS safe to me but just want to make sure both are intentional

Reply:
Yeah, intentional. I will clarify with a comment.

1. It is safe to call twice, even if it's not needed. So we call it eagerly to release the reflection stream.
2. If we need to unmarshal an Any or an extension in the JSON input, it will create a new reflection stream to resolve the descriptors for those. So the defer is necessary to make sure it shuts that stream down gracefully.

I mainly did this to make the output look cleaner if you use the `-v` flag. Without the first call, it's a bit confusing that reflection stream doesn't complete until _after_ the main RPC completes. However, if it does need to re-create a stream (like to unmarshal an unrecognized message in an Any or if it encounters an unknown extension), then that still happens. So maybe it's better to just always rely on defer (and then there are fewer reflection requests). WDYT?

Reply:
I removed the explicit close. Now it's only closed from the defer. Since I added the `[#2]` snippet to the request tracing, it's more clear that the final call completion is the reflection, so not as confusing as when I first added that call.